packages/expo-linking/src/Linking.ts

```diff
@@ -0,0 +1,334 @@
+  return url.replace(/\/?\?.*$/, '');
+}
+
+function ensureTrailingSlash(input: string, shouldAppend: boolean): string {
+  const hasSlash = input.endsWith('/');
+  if (hasSlash && !shouldAppend) {
+    return input.substring(0, input.length - 1);
+  } else if (!hasSlash && shouldAppend) {
+    return `${input}/`;
+  }
+  return input;
+}
+
+function ensureLeadingSlash(input: string, shouldAppend: boolean): string {
+  const hasSlash = input.startsWith('/');
+  if (hasSlash && !shouldAppend) {
+    return input.substring(1);
+  } else if (!hasSlash && shouldAppend) {
+    return `/${input}`;
+  }
+  return input;
+}
+
+/**
+ * Create a URL that works for the environment the app is currently running in.
+ * The scheme in bare and standalone must be defined in the app.json under `expo.scheme`.
+ *
+ * **Examples**
+ *
+ * - Bare, Standalone: yourscheme://
+ * - Web (dev): https://localhost:19006/
+ * - Web (prod): https://myapp.com/
+ * - Expo Client (dev): exp://128.0.0.1:19000/--/
+ * - Expo Client (prod): exp://exp.host/@yourname/your-app/
+ * - Expo Client (prod): exp://exp.host/@yourname/your-app/
+ *
+ * @param path addition path components to append to the base URL.
+ * @param queryParams An object of parameters that will be converted into a query string.
+ */
+export function makeUrl(path: string = '', queryParams: QueryParams = {}): string {
+  if (Platform.OS === 'web') {
+    if (!canUseDOM) return '';
+
+    const origin = ensureTrailingSlash(window.location.origin, false);
+    let queryString = qs.stringify(queryParams);
+    if (queryString) {
+      queryString = `?${queryString}`;
+    }
+
+    let outputPath = path;
+    if (outputPath) outputPath = ensureLeadingSlash(path, true);
+
+    return encodeURI(`${origin}${outputPath}${queryString}`);
+  }
+  let scheme = 'exp';
+  const manifestScheme = manifest.scheme ?? manifest?.detach?.scheme;
+
+  if (Constants.appOwnership === 'standalone' && manifestScheme) {
+    scheme = manifestScheme;
+  } else if (Constants.appOwnership === 'standalone' && !manifestScheme) {
+    throw new Error('Cannot make a deep link into a standalone app with no custom scheme defined');
+  } else if (Constants.appOwnership === 'expo' && !manifestScheme) {
+    console.warn(
+      'Linking requires that you provide a `scheme` in app.json for standalone apps - if it is left blank, your app may crash. The scheme does not apply to development in the Expo client but you should add it as soon as you start working with Linking to avoid creating a broken build. Add a `scheme` to silence this warning. Learn more about Linking at https://docs.expo.io/versions/latest/workflow/linking/'
+    );
+  } else if (!Constants.manifest) {
+    // Bare-workflow
+    if (!manifestScheme) {
+      throw new Error(
+        "Cannot determine what scheme to use when making a URL because the `expo.scheme` property is not defined in the project's app.json"
+      );
+    }
+    scheme = manifestScheme;
```

Comment:
would we ever actually reach this code? it seems like we wouldn't, given that `manifestScheme` is set from `Constants.manifest` and we only enter this block if `Constants.manifest` is falsey.

also, in bare workflow we can access `manifest` from `Updates.manifest`, but this is only available in release mode when using expo-updates.

i think it's not super clear what the behaviour here should be in the situation where there are multiple schemes. maybe we should have an option where users can pass in some scheme to use with `makeUrl` and if not specified then use the "first one". we'll need to have a native method exposed to get the schemes to do this.